src/main/java/sparksoniq/spark/iterator/flowr/OrderByClauseSparkIterator.java

```diff
@@ -17,46 +17,149 @@
  * Author: Stefan Irimescu
  *
  */
- package sparksoniq.spark.iterator.flowr;
+package sparksoniq.spark.iterator.flowr;
 
 import org.apache.spark.api.java.JavaPairRDD;
 import org.apache.spark.api.java.JavaRDD;
-import sparksoniq.jsoniq.compiler.translator.expr.flowr.FLWOR_CLAUSES;
+import sparksoniq.exceptions.IteratorFlowException;
+import sparksoniq.exceptions.NonAtomicKeyException;
+import sparksoniq.exceptions.SparksoniqRuntimeException;
+import sparksoniq.jsoniq.item.Item;
+import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
 import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
 import sparksoniq.jsoniq.runtime.tupleiterator.RuntimeTupleIterator;
+import sparksoniq.jsoniq.runtime.tupleiterator.SparkRuntimeTupleIterator;
 import sparksoniq.jsoniq.tuple.FlworKey;
 import sparksoniq.jsoniq.tuple.FlworTuple;
 import sparksoniq.semantics.DynamicContext;
 import sparksoniq.spark.closures.OrderByClauseSortClosure;
 import sparksoniq.spark.closures.OrderByMapToPairClosure;
-import sparksoniq.spark.iterator.flowr.base.FlowrClauseSparkIterator;
 import sparksoniq.spark.iterator.flowr.expression.OrderByClauseSparkIteratorExpression;
 
-import java.util.List;
+import java.util.*;
 
-public class OrderByClauseSparkIterator extends FlowrClauseSparkIterator {
+public class OrderByClauseSparkIterator extends SparkRuntimeTupleIterator {
     private final boolean _isStable;
     private final List<OrderByClauseSparkIteratorExpression> _expressions;
 
+    private List<FlworTuple> _localTupleResults;
+    private int _resultIndex;
+
     public OrderByClauseSparkIterator(RuntimeTupleIterator child, List<OrderByClauseSparkIteratorExpression> expressions,
                                       boolean stable, IteratorMetadata iteratorMetadata) {
-        super(child, null, FLWOR_CLAUSES.ORDER_BY, iteratorMetadata);
+        super(child, iteratorMetadata);
         this._expressions = expressions;
-        this._expressions.forEach(e -> this._children.add(e.getExpression()));
         this._isStable = stable;
+        _localTupleResults = new ArrayList<>();
+        _resultIndex = 0;
+    }
+
+    @Override
+    public boolean isRDD() {
+        return _child.isRDD();
+    }
+
+    @Override
+    public void open(DynamicContext context) {
+        super.open(context);
+
+        // isRDD checks omitted, as open is used for non-RDD(local) operations
+
+        if (this._child != null) {
+            _child.open(_currentDynamicContext);
+
+            setAllLocalResults();
```

Comment:
This means that all computation is essentially done on opening. For the local computation case this would not be terrible but it still breaks the iterator pattern.

Reply:
Stefan is correct that, normally, this should be done in the first next() call. I think we do have other places where this will need to be fixed. Can you open an issue for this?

Reply:
You're right, since we need to complete results over the entire set, I had mistakenly pushed this computation to the open. This has been changed with checking whether child iterator hasNext in open and triggering computation only in the next() now.

let for and where clauses were implemented calculating only a single element in the open() so I believe they comply with the pattern. I had the same misconception for the groupby clause which is also fixed now. Should I open an issue to make a full sweep later?

Reply:
Thanks! Yes, an issue is worth it. I think we have a few item iterators where we have materialization done in open(). Nothing urgent, though. Just for cleanliness.